### PR TITLE
[bitnami/mlflow] Add support for Google Cloud Storage as artifact store

### PR DESCRIPTION
**Pull request by mautini**
### Description of the change

This PR follows this one https://github.com/bitnami/charts/pull/26461 and integration of `google-cloud-storage` library in bitnami mlflow Docker image: https://github.com/bitnami/containers/pull/67246

It adds support to Google Cloud Storage as backend store for mlflow artifacts. The use could specify credentials via workload identity or application credential.

### Benefits

User will be able to use Google Storage as backend for artifact storage in mlflow tracking. This is useful for other that migrate from another mlflow instance (running with GCS as backend) or that cannot use other cloud provider.

### Possible drawbacks

More complexity to the helm chart / configuration? But it's a feature supported by mlflow so I think we have to expose it.

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

**Comment by andresbono** [comment]
Thank you @mautini for this contribution! The CI failed for some reason. Let me retry it and I will review the PR. Thanks!

File: bitnami/mlflow/templates/_helpers.tpl [hunk 1]
```diff
@@ -648,6 +654,33 @@ Return the S3 secret access key inside the secret
...
+{{- end -}}
+
+{{/*
+Return the GCS bucket
+*/}}
+{{- define "mlflow.v0.gcs.bucket" -}}
+    {{- if and (not .Values.minio.enabled) (not .Values.externalS3.host) .Values.externalGCS.bucket -}}
+        {{- print .Values.externalGCS.bucket -}}
+    {{- end -}}
+{{- end -}}
```
Comment (andresbono) [inline bitnami/mlflow/templates/_helpers.tpl:682]: I think this helper is not adding any value. We could just get the value from `.Values.externalGCS.bucket` as we are already wrapping the calls to the helper with the same condition, right?

```suggestion
```

I guess it's to follow the same approach than for `mlflow.v0.s3.bucket`, but in that case, the helper has a relevant if/else.

File: bitnami/mlflow/templates/tracking/deployment.yaml [hunk 2]
```diff
@@ -141,12 +141,17 @@ spec:
             {{- end }}
             {{- if (include "mlflow.v0.s3.enabled" .) }}
             - --artifacts-destination=s3://{{ include "mlflow.v0.s3.bucket" . }}
+            {{- else if (include "mlflow.v0.gcs.enabled" .) }}
+            - --artifacts-destination=gs://{{ include "mlflow.v0.gcs.bucket" . }}
```
Comment (andresbono) [inline bitnami/mlflow/templates/tracking/deployment.yaml:145]:
```suggestion
            - --artifacts-destination=gs://{{ .Values.externalGCS.bucket }}
```

File: bitnami/mlflow/templates/tracking/deployment.yaml [hunk 3]
```diff
@@ -141,12 +141,17 @@ spec:
...
             {{- else }}
             - --artifacts-destination={{ .Values.tracking.persistence.mountPath }}/mlartifacts
             {{- end }}
             {{- if and (not (include "mlflow.v0.s3.serveArtifacts" .)) (include "mlflow.v0.s3.enabled" .) }}
             - --default-artifact-root=s3://{{ include "mlflow.v0.s3.bucket" . }}
             - --no-serve-artifacts
+            {{- else if and (not .Values.externalGCS.serveArtifacts) (include "mlflow.v0.gcs.enabled" .) }}
+            - --default-artifact-root=gs://{{ include "mlflow.v0.gcs.bucket" . }}
```
Comment (andresbono) [inline bitnami/mlflow/templates/tracking/deployment.yaml:153]:
```suggestion
            - --default-artifact-root=gs://{{ .Values.externalGCS.bucket }}
```

File: bitnami/mlflow/templates/tracking/deployment.yaml [hunk 4]
```diff
@@ -273,6 +288,11 @@ spec:
               {{- if .Values.tracking.persistence.subPath }}
               subPath: {{ .Values.tracking.persistence.subPath }}
               {{- end }}
+            {{- if and (include "mlflow.v0.gcs.enabled" .) .Values.externalGCS.useCredentialsInSecret .Values.externalGCS.existingSecret }}
+            - name: google-application-credentials
```
Comment (andresbono) [inline bitnami/mlflow/templates/tracking/deployment.yaml:292]: There is no volume called like that, should it be `gcs-key`? Or the other way around.

File: bitnami/mlflow/templates/tracking/deployment.yaml [hunk 5]
```diff
@@ -273,6 +288,11 @@ spec:
...
+              mountPath: /bitnami/gcs/key.json
+              subPath: key.json
```
Comment (andresbono) [inline bitnami/mlflow/templates/tracking/deployment.yaml:294]: Is this combination of `mountPath` and `subPath` configuring the mount volume as expected?

File: bitnami/mlflow/values.yaml [hunk 6]
```diff
@@ -1421,3 +1422,20 @@ externalS3:
...
   bucket: "mlflow"
   serveArtifacts: true
+
+## @section External Google Cloud Storage parameters
+## All of these values are only used when minio.enabled is set to false and externalS3 is not configured (host is empty)
+## @param externalGCS.bucket GCS bucket name. Activate gcs artifact storage if set
+## @param externalGCS.googleCloudProject Google Cloud Project to use (optional, needed when using "default application credentials")
+## @param externalGCS.useCredentialsInSecret Whether to read the GCS application credentials from a secret
+## @param externalGCS.existingSecret Name of an existing secret key containing the application credentials file (used only if useCredentialsInSecret is true)
+## @param externalGCS.existingSecretKey Key in the existing secret containing the application credentials (must be set if existingSecret is set)
```
Comment (andresbono) [inline bitnami/mlflow/values.yaml:1432]: How does auth work if a user does not provide the credentials in a secret? Are these parameters actually optional?

**Review by andresbono — COMMENTED** [review]
Thank you for your contribution!! 🎉  I left some suggestions and questions. Could you please look at them?

Also, could you share if you performed some specific validation of the changes on your own env? To better understand the use case as well.

File: bitnami/mlflow/templates/tracking/deployment.yaml (hunk 4 above)
Comment (mautini) [reply bitnami/mlflow/templates/tracking/deployment.yaml:292]: You're right. I took this part from the original MR there: https://github.com/bitnami/charts/pull/26461 as I don't use this authentication method. Let's rename it to `gcs-key`

File: bitnami/mlflow/templates/tracking/deployment.yaml (hunk 5 above)
Comment (mautini) [reply bitnami/mlflow/templates/tracking/deployment.yaml:294]: I also took that from the original MR. I just checked, and there was a problem with the volume configuration. I push a commit. (`path: key.json` vs `path: /bitnami/gcs/key.json`)

I tested locally using minikube and the following commands (if you want to test on your side):

```shell
kubectl create namespace test-bitnami
kubectl -n test-bitnami create secret generic gcs-secret --from-file=gcs_keys=./keys.json
kubectl -n test-bitnami apply -f deployment.yaml
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wait-deployment
spec:
  replicas: 1
  selector:
    matchLabels:
      app: wait-app
  template:
    metadata:
      labels:
        app: wait-app
    spec:
      containers:
        - name: wait-container
          image: busybox
          command: ['sh', '-c', 'while true; do sleep 3600; done']
          ports:
            - containerPort: 80
          volumeMounts:
            - name: gcs-key
              mountPath: /bitnami/gcs/key.json
              subPath: key.json
      volumes:
        - name: gcs-key
          secret:
            secretName: gcs-secret
            items:
              - key: gcs_keys
                path: key.json
```

The `keys.json` is correctly mounted at `/bitnami/gcs/key.json` and this is the path specified for `GOOGLE_APPLICATION_CREDENTIALS` env variable.

**Comment by mautini** [comment]
Hi @andresbono, thanks a lot for your careful review. I applied your suggested changes, responded to your questions and pushed a fix for the mounting of `key.json`. Let me know if there is something left.

File: bitnami/mlflow/templates/_helpers.tpl [hunk 7]
```diff
@@ -648,6 +654,24 @@ Return the S3 secret access key inside the secret
...
+
+{{/*
+Return whether artifacts should be served from GCS
+*/}}
+{{- define "mlflow.v0.gcs.serveArtifacts" -}}
+    {{- if .Values.externalGCS.serveArtifacts  -}}
+        {{- true }}
+    {{- end -}}
+{{- end -}}
```
Comment (andresbono) [inline bitnami/mlflow/templates/_helpers.tpl:673]: Let's remove this helper too

```suggestion
```

File: bitnami/mlflow/templates/tracking/deployment.yaml [hunk 8]
```diff
@@ -310,4 +330,12 @@ spec:
         {{- if .Values.tracking.extraVolumes }}
         {{- include "common.tplvalues.render" (dict "value" .Values.tracking.extraVolumes "context" $) | nindent 8 }}
         {{- end }}
+        {{- if and (include "mlflow.v0.gcs.enabled" .) .Values.externalGCS.useCredentialsInSecret .Values.externalGCS.existingSecret }}
```
Comment (andresbono) [inline bitnami/mlflow/templates/tracking/deployment.yaml:333]:
```suggestion
        {{- if and (include "mlflow.v0.gcs.enabled" .) .Values.externalGCS.useCredentialsInSecret }}
```

File: bitnami/mlflow/templates/tracking/deployment.yaml [hunk 9]
```diff
@@ -273,6 +288,11 @@ spec:
               {{- if .Values.tracking.persistence.subPath }}
               subPath: {{ .Values.tracking.persistence.subPath }}
               {{- end }}
+            {{- if and (include "mlflow.v0.gcs.enabled" .) .Values.externalGCS.useCredentialsInSecret .Values.externalGCS.existingSecret }}
```
Comment (andresbono) [inline bitnami/mlflow/templates/tracking/deployment.yaml:291]:
```suggestion
            {{- if and (include "mlflow.v0.gcs.enabled" .) .Values.externalGCS.useCredentialsInSecret }}
```

File: bitnami/mlflow/values.yaml (hunk 6 above)
Comment (andresbono) [inline bitnami/mlflow/values.yaml:1432]:
```suggestion
## @param externalGCS.existingSecret Name of an existing secret key containing the application credentials file (required when useCredentialsInSecret is true)
## @param externalGCS.existingSecretKey Key in the existing secret containing the application credentials (required when useCredentialsInSecret is true)
```

File: bitnami/mlflow/templates/tracking/deployment.yaml (hunk 5 above)
Comment (andresbono) [reply bitnami/mlflow/templates/tracking/deployment.yaml:294]: Thank you for double-checking it!

File: bitnami/mlflow/values.yaml (hunk 6 above)
Comment (andresbono) [reply bitnami/mlflow/values.yaml:1432]: Oh that's nice. Thank you for providing the context!

**Review by andresbono — COMMENTED** [review]
Thank you for the additional context!

I just added a few minor suggestions. If you agree with them, you can apply them and the PR is ready to go 🚀

**Comment by mautini** [comment]
All good for me, thanks!

**Review by andresbono — APPROVED** [review]
LGTM, thank you so much!

**Pull request merged**